File: src/demo_cpp_package/src/add_two_ints_client.cpp
#include <thread>
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/srv/add_two_ints.hpp"

class AddTwoIntsClientNode : public rclcpp::Node
{
    using AddInts = example_interfaces::srv::AddTwoInts;

public:
    AddTwoIntsClientNode() : Node("add_two_ints_cclient")
    {
        RCLCPP_INFO(get_logger(), "Service client has started");
        thread_ = std::thread([this](){callAddTwoIntsService(5, 4);});
    }

    void callAddTwoIntsService(int a, int b) 
    {
        auto client = create_client<AddInts>("add_two_ints");
        while (!client->wait_for_service(std::chrono::seconds(1)))
        {
            RCLCPP_INFO(get_logger(), "Waiting for communication with server");
        }

        auto request = std::make_shared<AddInts::Request>();
        request->a = a;
        request->b = b;

        auto future = client->async_send_request(request);

        try 
        {
            auto response = future.get();
            RCLCPP_INFO(get_logger(), "%d + %d = %d", a, b, response->sum);
        }
        catch (const std::exception& e)
        {
            RCLCPP_ERROR(get_logger(), "Service call failed");
        }
    }
private:
    std::thread thread_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<AddTwoIntsClientNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: src/demo_python_package/demo_python_package/add_two_ints_client.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from functools import partial

from example_interfaces.srv import AddTwoInts

class AddTwoIntsClientNode(Node):
    def __init__(self):
        super().__init__("add_two_ints_client")
        self.callAddTwoIntsService(5, 3)

    def callAddTwoIntsService(self, a, b):
        client = self.create_client(AddTwoInts, "add_two_ints")
        while not client.wait_for_service(1.0):
            self.get_logger().info("Waiting for AddTwoIntsServer to connect")
        
        request = AddTwoInts.Request()
        request.a = a
        request.b = b

        future = client.call_async(request=request)
        future.add_done_callback(partial(self.addTwoIntsCallback, a=a, b=b))

    def addTwoIntsCallback(self, future, a, b):
        try:
            response = future.result()
            self.get_logger().info(str(a) + " + " + str(b) + " = " + str(response.sum))
        except Exception as e:
            self.get_logger().error(f"Service call failed {e}")

def main(args=None):
    rclpy.init(args=args)
    node = AddTwoIntsClientNode()
    rclpy.spin(node)
    rclpy.shutdown()

if __name__=="__main__":
    main()